Balanced node-wide throughput limits

dlex opened the pull request:
Node-wide throughput quotas introduced in #8023 are now balanced between the shards withing each node. New config properties added to control the balancer and minimum amount of quotas in each shard. All properties are updateable at runtime. Included is refactoring of #8023 that makes shard-(and-higher)-level quota manager an independent sharded service.

Fixes #7855.

<!--

See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md##pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED.

Describe, in plain language, the motivation behind the change (bug fix,
feature, improvement) in this PR and how the included commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.

  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.

  Backport of PR #PR-NUMBER

-->

## Backports Required

<!--

Checking at least one of the checkboxes is REQUIRED if this PR is not a backport.

-->

- [x] none - not a bug fix
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v22.3.x
- [ ] v22.2.x
- [ ] v22.1.x

## UX Changes

New cluster configuration properties:
| Name | Description |
| --- | --- |
| kafka_quota_balancer_node_period_ms | Intra-node throughput quota balancer invocation period, in milliseconds. Value of 0 disables the balancer and makes all the thoughput quotas immutable. |
| kafka_quota_balancer_min_shard_thoughput_ratio | The lowest value of the throughput quota a shard can get in the process of quota balancing, expressed as a ratio of default shard quota. 0 means there is no minimum, 1 means no quota can be taken away by the balancer. |
| kafka_quota_balancer_min_shard_thoughput_bps | The lowest value of the throughput quota a shard can get in the process of quota balancing, in bytes/s. 0 means there is no minimum. |


<!--

Content in this section is OPTIONAL.

Describe, in plain language, how this PR affects an end-user. Explain
topic flags, configuration flags, command line flags, deprecation
policies, etc. that are added or modified. Don't ship user breaking
changes. Ask the @redpanda-data/product team if you need help with user
visible changes.

-->

## Release Notes

<!--

Adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

  ### Bug Fixes

  * Short description of the bug fix if this is a PR to `dev` branch.
-->
  ### Features

  * Throughput limit can now be applied at node scope. Each individual node will limit the overall ingress throughput of all Kafka client connections it serves to the number specified by `kafka_throughput_limit_node_in_bps` cluster property; same can be done on the egress side using `kafka_throughput_limit_node_out_bps`.

<!--
  ### Improvements

  * Short description of how this PR improves existing behavior.

If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

  * none

-->

dlex commented:
Force push: lint errors fixed

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -166,6 +203,52 @@ snc_quota_manager::get_quota_balancer_node_period() const {
+}
+
+void snc_quota_manager::update_shard_quota_minimum() {
+    _shard_quota_minimum = to_each(
+      [this](const std::optional<quota_t>& node_quota_default) {
+          const auto shard_quota_default = node_to_shard_quota(
+            node_quota_default);
+          return std::max<quota_t>(
+            _kafka_quota_balancer_min_shard_thoughput_ratio()
+              * shard_quota_default,
+            _kafka_quota_balancer_min_shard_thoughput_bps());
+      },
+      _node_quota_default);
+    // downstream updates: none
```
`to_each` is still unintuitive to me. is this effectively the same thing as:

```c++
auto f = [this](p) {
          const auto shard_quota_default = node_to_shard_quota(p);
          return std::max<quota_t>(
            _kafka_quota_balancer_min_shard_thoughput_ratio()
              * shard_quota_default,
            _kafka_quota_balancer_min_shard_thoughput_bps());
};
_shard_quota_minimum.in = f(_node_quota_default.in);
_shard_quota_minimum.out = f(_node_quota_default.out);
```

dlex replied:
Exactly. The name `to_each` is a shorthand from "apply the function to each side of every argument, and combine the results".

dotnwat replied:
it kind of seems like this is a perfect case to use c++ ranges or boost ranges. almost all your ranges would have length 2, but a few cases it looks like you might have range lenght 4.

dlex replied:
refactored

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -310,8 +315,212 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+           == collect_dispense_amount_t::value) {
+        co_await container().invoke_on_all([shard_quotas_update](
+                                             snc_quota_manager& qm) {
+            to_each(
+              [](
+                bottomless_token_bucket& shard_quota,
+                const collect_dispense_amount_t& shard_quotas_update) {
+                  if (
+                    shard_quotas_update.amount_is
+                    == collect_dispense_amount_t::value) {
+                      shard_quota.set_quota(shard_quotas_update.amount);
+                  }
+              },
+              qm._shard_quota,
+              shard_quotas_update);
```
this is confusing because the return value of to_each is thrown away and it seems you only care about the side effect of the lambda that is applied. in this case it seems like it would be far simpler to have a lambda function and call it on shard_quota.{in,out} etc...?

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -310,8 +315,212 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+              });
+            to_each(
+              [](
+                std::vector<quota_t>& schedule,
+                quota_t& delta,
+                const quota_t total_surplus,
+                const std::vector<quota_t>& surplus) {
+                  if (delta < 0) {
+                      if (delta > -total_surplus) {
+                          quota_t remainder = delta;
+                          // pro rata to surpluses
+                          for (size_t k = 0; k != ss::smp::count; ++k) {
+                              const quota_t share = muldiv(
+                                delta, surplus.at(k), total_surplus);
+                              schedule.at(k) += share;
+                              remainder -= share;
+                          }
+                          // the remainder equally
+                          vassert(
+                            remainder <= 0,
+                            "Expected {} <= 0; delta: {}, total_surplus: {}, "
+                            "surpluses: {}",
+                            remainder,
+                            delta,
+                            total_surplus,
+                            surplus);
+                          if (remainder < 0) {
+                              const quota_t d = (remainder + 1)
+                                                  / static_cast<quota_t>(
+                                                    schedule.size())
+                                                - 1;
+                              for (quota_t& s : schedule) {
+                                  const quota_t dd = std::max(d, remainder);
+                                  remainder -= dd;
+                                  s += dd;
+                              }
+                          }
+                      } else { // delta <= -total_surplus
+
+                          // all surpluses
```
factor out into helper functions

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -310,8 +315,212 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+                          }
+                      }
+                  }
+              },
+              schedule,
+              deltas,
+              total_surplus,
+              to_inside_out(surplus));
+        }
+        if (deltas.in > 0 || deltas.out > 0) {
+            to_each(
+              [](std::vector<quota_t>& schedule, quota_t& delta) {
+                  if (delta > 0) {
+                      // equally
+                      auto share = std::div(delta, ss::smp::count);
+                      for (quota_t& s : schedule) {
+                          s += share.quot;
+                          if (share.rem > 0) {
+                              s += 1;
+                              share.quot -= 1;
+                          }
+                      }
+                  }
+              },
+              schedule,
+              deltas);
+        }
```
definitely needs explicit comments about what is happening.

dlex replied:
I think this code is self doc now

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -310,8 +315,212 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
 }
 
 ss::future<> snc_quota_manager::quota_balancer_update() {
```
this needs a comment above that explain what exactly is happening here.

dlex replied:
a comment is added in the header - pls let me know if that is not what you've proposed

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -309,9 +310,95 @@ void snc_quota_manager::quota_balancer() {
+          return to_collect;
+      },
+      inoutpair<quota_t>{0, 0},
+      [](const inoutpair<quota_t>& s, const inoutpair<quota_t>& v) {
+          return to_each(std::plus{}, s, v);
+      });
+    vlog(klog.trace, "qb - Collected: {}", collected);
+
+    // dispense the collected amount among the borrowers
+    auto split = to_each(
+      [](const quota_t collected, const shard_count_t borrowers_count) {
+          if (borrowers_count == 0) {
+              vassert(
+                collected == 0, "Quota collected when borrowers are absent");
+              return std::div(collected, shard_count_t{1});
+          }
+          return std::div(collected, borrowers_count);
+      },
+      collected,
+      borrowers_count);
+    for (const borrower_t& b : borrowers) {
+        if (b.is_borrower.in == 0 && b.is_borrower.out == 0) {
+            continue;
+        }
+        const inoutpair<quota_t> share = to_each(
+          [](auto& split, const shard_count_t is_borrower) -> quota_t {
+              if (is_borrower == 0) {
+                  return 0;
+              }
+              if (split.rem == 0) {
+                  return split.quot;
+              }
+              --split.rem;
+              return split.quot + 1;
+          },
+          split,
+          b.is_borrower);
+        // TBD: make the invocation parallel
+        co_await container().invoke_on(
+          b.shard_id,
```
are there some distinct components here that could be factored out into separate functions and explain indivdiually? ok if there is not.

dotnwat commented on src/v/kafka/server/tests/produce_consume_test.cc:
```diff
@@ -305,13 +353,85 @@ FIXTURE_TEST(test_node_throughput_limits, prod_consume_fixture) {
+    BOOST_REQUIRE_GT(kafka_in_data_len, kafka_out_data_len);
+}
+
+FIXTURE_TEST(test_node_throughput_limits_balanced, throughput_limits_fixure) {
```
have you thought about ducktape tests? ultimately that is what we are going to need.

dlex replied:
Definitely, I'm just not ready to PR them yet. The UTs are basically the bare minimum.

dotnwat commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -19,10 +19,84 @@
+  class... T,
+  std::invocable<detail::unwrap_inoutpair_t<T>...> F,
+  class TRes = std::invoke_result_t<F, detail::unwrap_inoutpair_t<T>...>>
+inoutpair<TRes> to_each(F f, T&&... args) {
+    if constexpr (std::is_void_v<TRes>) {
+        std::invoke(f, args.in...);
+        std::invoke(f, args.out...);
+        return {};
```
oh i see there is no recursion.

i think what you want is two functions:

* apply
* reduce

dotnwat commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -19,10 +19,84 @@
+
+/// Copies vector of inoutpairs to an inoutpair of vectors
+template<class T>
+inoutpair<std::vector<T>> to_inside_out(const std::vector<inoutpair<T>>& v) {
```
there is only one caller to this it might make sense to perform this inline where it's called because the name isn't very intuitive about what is happening, but the actual implementation makes it clear.

dlex replied:
kept the function but moved it to the .cc

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -310,8 +315,212 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+          }
+          return quota_t{0};
+      },
+      shard_quotas_update);
```
is this something like?

```
f = [] { ... };
inoutpair<quota_t> deltas;
deltas.in = f(shard_quotas_update.in);
deltas.out = f(shard_quotas_update.out);
```

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -310,8 +315,212 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+            const auto total_quota = co_await container().map_reduce0(
+              [](const snc_quota_manager& qm) {
+                  return to_each(
+                    std::minus{}, qm.get_quota(), qm._shard_quota_minimum);
```
is this the same thing?

```c++
return inoutpair{
  .in = qm.get_quota().in - qm._shard_quota_minimum.in,
  .out = qm.get_quota().out - qm._shard_quota_minimum.out
};
```

graphcareful reviewed (changes requested):
I'm pretty much in alignment with Noah's comments so I will not leave duplicate comments on whats already been reviewed. However i'll state my bigger opinions that haven't already been addressed here.

It was kind of hard to review, I think breaking things up into smaller named functions would definitely be the way to go. There were some pretty large lambdas passed into `to_each` in parts of the core algorithm too. Breaking these out into smaller units would also mean you could unit test the pieces individually.

I see what is going on with `inoutpair<T>`, however I don't understand the necessity. The in and out data are always kept separate and never aggregated. Which leads me to believe the algorithm can just be run twice, once on the `in` data and again on the `out`. I guess you wanted to avoid this because that would mean double the async work dispatched, but this can easily be avoided if things were broken up into smaller units. Example:
```
f = (int a) { .... };
auto new_ingress = f(ingress);
auto new_egress = f(egress);
auto per_shard_resp = co_await container().map([new_ingress, new_egress](snd_manager& s){
    ... continue algorithm... 
    return std::make_tuple(per_shard_work(new_ingress), per_shard_work(new_egress));
});
// continue, calling pieces of the algorithm twice once on ingress then egress.
```

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -138,3 +140,18 @@ void snc_quota_manager::record_response_tp(
 
 } // namespace kafka
+
+struct parseless_formatter {
```
To avoid defining dummy `parse()` funcs in each formatter specialization.

dotnwat replied:
i dunno about this `fmt` stuff... seems like `operator<<` is simpler. maybe there are ways to manage the complexity.

nothing to change here, i think we have a tree-wide thing to understand about fmt.

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -50,11 +55,28 @@ snc_quota_manager::snc_quota_manager()
-ss::future<> snc_quota_manager::start() { return ss::make_ready_future<>(); }
+ss::future<> snc_quota_manager::start() {
+    if (ss::this_shard_id() == quota_balancer_shard) {
+        _balancer_thread = ss::thread([this] { quota_balancer(); });
```
that would be nice, but it looks like this will be ok. we can "modernize" and move to coroutines as a follow up, i think.

dlex replied:
yep refactored into the timer based approach now

dotnwat commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -0,0 +1,82 @@
+
+namespace kafka {
+
+class snc_quota_manager
```
agree. lots cleaner than with the sharded<x>* _container;

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -127,6 +148,89 @@ void snc_quota_manager::record_response_tp(
+    _as.request_abort_ex(abort_on_configuration_change{});
+    // replace abort source immediately so that only subscribers are aborted now
+    // and to enable future aborts
+    _as = ss::abort_source();
```
refactored

dlex commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -136,8 +137,36 @@ class snc_quota_manager
+        bool empty() const noexcept {
+            return amount == 0 && amount_is == delta;
+        }
+        void append(const collect_dispense_amount_t& rhs) noexcept;
```
gone in refactoring

dlex commented:
Force push:
* refactoring of the excessively vectorized `inoutpair` algorithms, code structure split into smaller functions, comments and readability
* `ss::thread` discarded in favour of timer+gate+semaphore approach as discussed later, massive refactoring as the result with multiple simplifications
* other review feedback addressed, although not everything yet

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -163,6 +185,51 @@ void snc_quota_manager::record_response_tp(
+}
+
+ss::future<> snc_quota_manager::quota_balancer_step() {
+    const auto sem_units = co_await ss::get_units(_balancer_semaphore, 1);
```
I see you've added this here for later use. I spent a few minutes scrutinizing this before I saw how it was used in future commits. I would in this commit place a comment like "for reviewers , inserting scaffolding , async work to be done here" that eventually will be deleted on subsequent commits when you do introduce that logic.

dotnwat replied:
agree. or move its introduction to a later commit where it is needed?

dlex replied:
sometimes it's hard to reason about the features when there is also a time dimension in the codebase :)

graphcareful reviewed (commented):
Nice improvements!

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -13,10 +13,50 @@
 namespace kafka {
 
+template<std::integral T>
+inoutpair<T> operator+(const inoutpair<T>& lhs, const inoutpair<T>& rhs) {
```
I would rename this struct `ingress_egress_stats`. `inoutpair` is not immediately obvious to code readers whats the purpose of the struct

dlex replied:
it works with more than just `quota_t` so it's not always a "quota"; "stats" looks confusing to me as well. I agree the name is somewhat unclear though... some ideas may be
* `ingress_egress` - then the member names should be changed too IMO
* 'in_out` - would that be good enough?

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -481,6 +575,20 @@ void snc_quota_manager::refill_buckets(const clock::time_point now) noexcept {
 }
 
+inoutpair<quota_t> snc_quota_manager::get_deficiency() const noexcept {
+    const auto f = [](
```
can be marked `static const`

dotnwat replied:
what would marking `f` as `static` achieve? it doesn't seem to have any state, for instance.

graphcareful replied:
I guess nothing other then avoiding the initialization of `f` once per invocation. Logically I see it as a non-changing constant for the duration of the program, I generally mark things like that as `static + const`. Effectively, its the same as just making this lambda a function.

dotnwat replied:
> I guess nothing other then avoiding the initialization of f once per invocation
i think we are probably underestimating the compiler if we believe that it isn't inline this.

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -297,10 +298,103 @@ void snc_quota_manager::maybe_arm_balancer_timer() {
+struct borrower_t {
+    ss::shard_id shard_id;
+    // can be 0 or 1, made shard_count_t for easier aggregation
+    inoutpair<shard_count_t> is_borrower;
```
The comment and data type aren't clear. `is_borrower` makes me think this can be 0/1 (aligned with the comment) but the data type is `shard_count_t` making me think its a number of shards which are borrowers?

dotnwat replied:
yeh, seems like this could be inoutpair<bool>?

dotnwat replied:
it's generally better to have types that make sense. why wouldn't you use convert is_borrower to a number in the lambda here? otoh, if it is a count, then name it that.

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -297,10 +298,103 @@ void snc_quota_manager::maybe_arm_balancer_timer() {
+
+    // determine the borrowers and whether any balancing is needed now
+    const auto now = clock::now();
+    const std::vector<borrower_t> borrowers = co_await container().map(
```
I think this + counting borrowers can be a function named `total_borrowers_count` or something like that

dlex replied:
did you actually mean the `+` in the `accumulate` context 10 lines down?

graphcareful replied:
the context 10 lines down, yeah

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -303,10 +303,228 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+          shard_quotas_update,
+          schedule);
+
+        co_await container().invoke_on_all([&schedule](snc_quota_manager& qm) {
```
Looks nice how everything boils down to this call now +1

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -163,6 +185,51 @@ void snc_quota_manager::record_response_tp(
+    const auto now = ss::steady_clock_type::now();
+    const ss::steady_clock_type::time_point closest_arm_until
+      = now + min_sleep_time;
+    if (arm_until < closest_arm_until) {
```
> I think this should be refactored to just run on the configured period (its a bounded property that has a static min right?). So we shouldn't have to worry about the case where the loop runs too tight because the min sleep will bound that.

Static min is not the same: intervals are measured at the beginning of the balancer processing so that the time that it takes for a balancer step to run does not influence its frequency. So even with a static minimum it is possible that (run time of the balancer step) > (balancer period) => no delay between balancer runs. This is when this code is supposed to intervene to prevent the balancer from taking too much share of the core0 compute.

dlex replied:
> Furthermore I don't think its an issue if the algorithm takes longer to run then expected and therefore doesn't run every Xms

I was thinking about high load scenarios. If the reactor is really busy, it can take signinficantly longer for cross shard calls to complete. If we don't try to keep the balancer frequency stable, this may cause the situations like "balancer stops working under high load" (actully it doesn't but only works much slower) -- while the high load scenarios is where it is expected to work in first place.

The next step to make balancer less sensitive to high load is to allow multiple balancer steps to run in parallel, but I have left that improvement for later.

dotnwat replied:
there should be no assumptions about _when_ code is going to be executed.

dlex replied:
> there should be no assumptions about _when_ code is going to be executed

Could you please elaborate?

dotnwat replied:
correctness shouldn't be predicated on the timer running at a precise time

graphcareful replied:
Yeah i'm still kind of on the fence about this one. To me it makes sense that the only thing we should guarantee, is that invocations of the algorithm occur at least `interval_ms` between each other.

In general (with this fiber/timer design pattern) do we ever care when the time it takes for one iteration of a fiber exceeds its periodicity interval? I don't think its an explicitly handled condition in other places of the codebase today @dotnwat am I right in assuming this?

I think another pattern to consider might be to let the loop run at the configured interval, but maybe have the loop just do no work and return if its observed this condition occur.

dlex replied:
> correctness shouldn't be predicated on the timer running at a precise time

> To me it makes sense that the only thing we should guarantee, is that invocations of the algorithm occur at least `interval_ms` between each other.

In this case it's not _correctness_, it's _precision_ of the implemented controls. If balancer frequency fluctuates because of the time it takes for an iteration to complete, that does not render the balancer effort incorrent, but the response time/magnitude will be somewhat (slightly or not) different from what may be expected from the current configuration.

dlex commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -83,6 +112,12 @@ class snc_quota_manager
+    ss::timer<ss::steady_clock_type> _balancer_timer;
+    ss::steady_clock_type::time_point _balancer_timer_last_ran;
+    ss::gate _balancer_gate;
+    ss::semaphore _balancer_semaphore{1};
```
> I don't think the semaphore is necessary because you are controlling the execution of the fiber and you can make assumptions of concurrency. If you only call `arm` once every iteration, then there can't be a case where multiple invocations of the timer cause two fibers to run the algorithm at the same time. So i think this can just be removed.

That's about balancer stability and tolerance to changing load on the reactor: see https://github.com/redpanda-data/redpanda/pull/8335#discussion_r1084358184

dlex replied:
You're actually right about `quota_balancer_step()` alone: the timer is rearmed after each step is executed, so there should be no implication on concrrency. The semaphore is needed here to prevent concrrency between `quota_balancer_step()` and `quota_balancer_update()`, as well as between potentially concurrent fibers of the latter.

dlex replied:
> If you did need a semaphore of 1 unit check out `ssx::mutex` instead

I see why I could not find it: it's in `utlis/` not in `ssx/`. However `ssx/` would have been a better fit wouldn't it?

dlex commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -135,25 +136,21 @@ class snc_quota_manager
     // operational, used on each shard
-    bottomless_token_bucket _shard_ingress_quota;
-    bottomless_token_bucket _shard_egress_quota;
+    inoutpair<bottomless_token_bucket> _shard_quota;
```
that's a way to express the fact that this setting is also symmetrinc for ingress/egress. From this perspective, changing the wrapper name to `ingress_egress` looks more reasonable.

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -303,10 +303,228 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+    vlog(klog.trace, "qb - Update: {}", shard_quotas_update);
+
+    // deliver shard quota updates of value type
+    if (
+      shard_quotas_update.in.amount_is == collect_dispense_amount_t::value
+      || shard_quotas_update.out.amount_is
+           == collect_dispense_amount_t::value) {
+        co_await container().invoke_on_all(
+          [shard_quotas_update](snc_quota_manager& qm) {
+              qm.maybe_set_quota(
+                {.in = shard_quotas_update.in.get_value(),
+                 .out = shard_quotas_update.out.get_value()});
+          });
+    }
```
much easier to read!

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -303,10 +303,228 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+            qm.adjust_quota(
+              {.in = schedule.in.at(ss::this_shard_id()),
+               .out = schedule.out.at(ss::this_shard_id())});
+            vlog(klog.trace, "qb - Delta updates applied: {}", qm._shard_quota);
```
this looks great

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -297,10 +298,103 @@ void snc_quota_manager::maybe_arm_balancer_timer() {
     _balancer_timer.arm(arm_until);
 }
 
+using shard_count_t
+  = boost::uint_t<std::numeric_limits<ss::shard_id>::digits>::fast;
+
+struct borrower_t {
+    ss::shard_id shard_id;
+    // can be 0 or 1, made shard_count_t for easier aggregation
+    inoutpair<shard_count_t> is_borrower;
```
what is `boost::uint_t<std::numeric_limits<ss::shard_id>::digits>::fast;`? i think we should try to remove this. based solely on the type name `shard_count` an `unsigned` should work if you need it to match ss::shard_id.

dlex replied:
that basically means "unsigned type that is capable to hold `ss::shard_id`". I didn't want to use `ss::shard_id` as a shard count (it may be the same type but the domain is different), nor I wanted to hardcode a type based on what `shard_id` currently is. But maybe you're right that's an overkill and hardcoding that is good enough.

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -297,10 +298,103 @@ void snc_quota_manager::maybe_arm_balancer_timer() {
+
+    // determine the borrowers and whether any balancing is needed now
+    const auto now = clock::now();
+    const std::vector<borrower_t> borrowers = co_await container().map(
```
`const auto .. = `?

dlex replied:
If everyone is aware that `map()` retuns a `vector` and not any other container, that may be ok. I'm trying to resort to `auto` only when it's crystal clear from the nearby code what the type is ... and in templates :)

dotnwat replied:
i think it's safe to say that map returning a vector is common information.

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -297,10 +298,103 @@ void snc_quota_manager::maybe_arm_balancer_timer() {
+        split.out = std::ldiv(collected.out, borrowers_count.out);
+    }
+
+    const auto equal_share_or_0 =
+      [](split_t& split, const shard_count_t is_borrower) -> quota_t {
+        if (is_borrower == 0) {
+            return 0;
+        }
+        if (split.rem == 0) {
+            return split.quot;
+        }
+        --split.rem;
+        return split.quot + 1;
+    };
```
all of `quota_balancer_step` needs to have block comments injected to walk the reader through about what is happening at each step along the way.

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -297,10 +298,103 @@ void snc_quota_manager::maybe_arm_balancer_timer() {
+              [share](snc_quota_manager& qm) { qm.adjust_quota(share); });
+        }
+    }
+    vassert(
+      split.in.rem == 0 && split.out.rem == 0,
```
why is this guaranteed to be true? the consequences are a hard-stop of the system. is there a less severe way to handle this scenario?

dlex replied:
I think I will spill the ERROR into the log instead.

dotnwat replied:
sounds good

dotnwat commented:
Can you get rid of the semaphore? When update() method runs could you update data structures and state of the system but not actually kick off any async work, then rearm the timer so that the step() function runs to perform the heavy weight work, without anything running concurrently, and then you don't need the semaphore?

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -303,10 +303,228 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+        for (size_t k = 0; k != ss::smp::count; ++k) {
+            const quota_t share = muldiv(delta, surplus.at(k), total_surplus);
+            schedule.at(k) += share;
+            remainder -= share;
+        }
+        // the remainder equally
+        vassert(
+          remainder <= 0,
```
I was thinking the same too. Crashing the process is definitely a little too extreme, maybe we could keep a counter on these failures and shutdown the fiber upon observation of the number of these types of failures passing some static threshold.

dotnwat replied:
yeh. we could probably even avoid shutting down the fiber as long as (1) the exceptional case skipped applying any updates (or alternative, reset updates to some safe defaults) and (2) the exception case always maintained a paced execution (ie no cases where we'd spin).

if you feel confident in the assert, then it might still be useful to _not_ assert, but rather model it as an assertion failure: tear down all your quota rebalancing fibers, reset state, etc... and then internally restart the balancer.

dlex replied:
I was a bit lazy to formally prove that `remainder <= 0` will always hold, hence the vassert. I think I should do that and just remove the `vassert`.

dotnwat replied:
logging and error and dealing with the issue gracefully is probably sufficient for now. but a proof also works. the important thing is just that we don't hard stop redpanda, and that isn't obviously the case here.

dlex replied:
changed to error log

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -0,0 +1,133 @@
+  , _shard_ingress_quota(
+      get_shard_ingress_quota_default(), _kafka_quota_balancer_window())
+  , _shard_egress_quota(
+      get_shard_egress_quota_default(), _kafka_quota_balancer_window()) {
```
The container for "quota" is a token bucket:
```
inoutpair<bottomless_token_bucket> _shard_quota;
```

That was originally done to support the naming pattern in `quota_manager` where token buckets and rate trackers are all part of an entity called "quotas" (now `_client_quotas`).

dotnwat replied:
ok!

dlex commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -0,0 +1,82 @@
+    /// @p request delay to request from the client via throttle_ms
+    struct delays_t {
+        clock::duration enforce{0};
+        clock::duration request{0};
```
Yes it does integrate. Same as per-client throttling, the delay is sent in the response first, and is enforced in the next call if needed.

dlex commented on src/v/config/configuration.h:
```diff
@@ -352,6 +352,10 @@ struct configuration final : public config_store {
     bounded_property<std::chrono::milliseconds> kafka_quota_balancer_window;
+    bounded_property<std::chrono::milliseconds>
+      kafka_quota_balancer_node_period;
+    property<double> kafka_quota_balancer_min_shard_thoughput_ratio;
```
`bounded_property` requires an integral parameter type because of the `align` bound which requires operator `%` to be applicable to the type.

dotnwat replied:
you could probably if constexpr that out if it isn't an integral type and create a specialized version of numeric_bounds that could be used for floats.

but no need to change that in this PR.

dlex commented on src/v/config/configuration.cc:
```diff
@@ -1662,7 +1662,34 @@ configuration::configuration()
+  , kafka_quota_balancer_min_shard_thoughput_ratio(
+      *this,
+      "kafka_quota_balancer_min_shard_thoughput_ratio",
+      "The lowest value of the throughput quota a shard can get in the process "
+      "of quota balancing, expressed as a ratio of default shard quota. "
+      "0 means there is no minimum, 1 means no quota can be taken away "
+      "by the balancer.",
```
I actually think this setting is more useful than the absolute bps value, because it basically says, e.g. "5% of the shard's equal share of the node wide quota is always assigned to each shard no matter what". It's easier to reason about than the absolute value: we can see then that 95% can be taken away, which means that in an extreme case one shard can get up to 95% of the node wide limit.

But ultimately it's about convenience: since the strongest of the ratio and the absolute value is applied, it's up to the user which one to use.

dotnwat replied:
thanks!

since this doesn't fit into the configuration description, please add it to the commit message. you can write as much as you need to in the commit message to document stuff like this.

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -59,16 +59,13 @@ ss::future<> snc_quota_manager::stop() { return ss::make_ready_future<>(); }
+  const config::binding<std::optional<int64_t>>& prop) {
     if (prop()) {
-        const uint64_t v = *prop() / ss::smp::count;
-        if (v >= static_cast<uint64_t>(bottomless_token_bucket::max_quota)) {
```
There is no change in the type anymore, both are signed.

dotnwat replied:
what i'm asking is: if prop/cpu-count could have been over max_quota, what about a signed verison means that that is no longer possible?

dlex replied:
heck you're right!

dlex commented on src/v/config/configuration.cc:
```diff
@@ -1662,7 +1662,34 @@ configuration::configuration()
+      "kafka_quota_balancer_node_period_ms",
+      "Intra-node throughput quota balancer invocation period, in "
+      "milliseconds. Value of 0 disables the balancer and makes all the "
+      "thoughput quotas immutable.",
```
Well those constants are per-shard ingress and egress quotas, currently they are based on `kafka_throughput_limit_node_in_bps` and `kafka_throughput_limit_node_out_bps` most of the time, although there is a corner case, when this property is changed on the fly, the value of the quotas are whatever the balancer has set before that. Moreover, once there is cluster throughput limit, those constants are also influenced by the cluster wide setting and the number of nodes in the cluster (in future, not in this iteration).

I'm not sure it makes sense to try squeezing that info into configuration property description -- please confirm that it does and I will change it.

dotnwat replied:
Thanks for clarifying. Before thinking about the wording, I'm interested in the functionality which I didn't realise existed.

Functionality i'm not sure it makes sense to leave the values fixed at whatever value they are at when the balancer is disabled. From a usability perspective I don't think I'd disable the balancer unless something bad was happening, and in that case, it might be because the balancer was misbehaving.

In this case I'd probably expect that when the balancer was disabled that the quotas reset to something sane.

wdyt?

dlex replied:
> In this case I'd probably expect that when the balancer was disabled that the quotas reset to something sane.

That makes sense. So do you suggest when we see the change of this property to `0`, we should reset quotas to the default values as well? That may make sense, however that may look like an unexpected side effect though. 

Currently the way to manually reset the quotas is to disable limiting by removing the `kafka_throughput_limit_node_*_bps` property and then by putting it back. It's not the most convenient way to do that.

I vote for that feature.
- do you agree?
- ok for a later PR?

dotnwat replied:
> disable limiting by removing the kafka_throughput_limit_node_*_bps property and then by putting it back

i'm not sure what that means. you'd have to change the config option and then change it again? that doesn't seem intuitive.

> ok for a later PR?

yes, please create a ticket to follow this up. it's important.

dotnwat replied:
https://github.com/redpanda-data/redpanda/issues/8392

dlex replied:
thank you!

dlex commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -68,9 +68,8 @@ class snc_quota_manager
-      _kafka_throughput_limit_node_in_bps;
-    config::binding<std::optional<uint64_t>>
+    config::binding<std::optional<quota_t>> _kafka_throughput_limit_node_in_bps;
+    config::binding<std::optional<quota_t>>
```
My idea was that it would ensure that `quota_t` is the same type as the one of configuration property (otherwise compilation will fail), but without the need to mention `quota_t` in configuration.h.

dotnwat replied:
>My idea was that it would ensure that quota_t is the same type as the one of configuration property

for that you'd need a strong type.

dlex replied:
actually not: simply changing the type to unsigned gives
```
error: no viable conversion from 'binding<optional<long>>' to 'binding<optional<unsigned long>>'
```

dotnwat replied:
that's not a strong type, tho--you can change your typedef to match. binding doesn't hold a quota.

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -13,10 +13,50 @@
+inoutpair<T> operator-(const inoutpair<T>& v) {
+    return {.in = -v.in, .out = -v.out};
+}
+template<std::integral T>
+bool operator<(const inoutpair<T>& lhs, const inoutpair<T>& rhs) {
+    if (lhs.in < rhs.in) {
+        return true;
+    } else {
+        return lhs.out < rhs.out;
+    }
+}
```
On one hand you're right this ordering is used only once in the code. On the other one, I can't see anything stange in defining ordering for two independent variables of the same type, particularilly if the ordering is intuitive for them and actually reflects what you would get from std::pair/tuple in the same context.

dotnwat replied:
i just don't understand what it means for a tuple of things related to ingress/egress to have any ordering. one option to make it less confusing is to place a comment above it with an example of how it is used and why that is useful.

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -136,3 +176,18 @@ void snc_quota_manager::record_response_tp(
 } // namespace kafka
+
+struct parseless_formatter {
+    constexpr auto parse(fmt::format_parse_context& ctx)
```
4 overall, but that's not the issue. I feel like
```
struct fmt::formatter<kafka::inoutpair<T>> : parseless_formatter
```
names what the formatter is without the extra 4 lines, and that is much easier to read.

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -19,6 +19,8 @@ using namespace std::chrono_literals;
 
 namespace kafka {
 
+using quota_t = snc_quota_manager::quota_t;
```
Closer to 15 places actually, and only a few of them are return types of member functions.

I have decided that it makes sense because `quota_t` makes sense for the enire unit and this definition causes no ambiguity in its scope.

dotnwat replied:
except for the return values, it seems like most quota_t has a type definition that is available in scope because it already defines on the snc_quota_manager::quota_t.

dlex replied:
> except for the return values, it seems like most quota_t has a type definition that is available in scope because it already defines on the snc_quota_manager::quota_t.

there are many namespace functions dealing with it too

dotnwat replied:
ok

dlex commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -69,25 +70,21 @@ class snc_quota_manager
-    config::binding<std::optional<quota_t>> _kafka_throughput_limit_node_in_bps;
-    config::binding<std::optional<quota_t>>
-      _kafka_throughput_limit_node_out_bps;
+    inoutpair<config::binding<std::optional<quota_t>>>
```
It does, the "limit" kind of property defines maximum bound for quotas, so its type is the type of quotas.

dotnwat replied:
it holds an optional int64_t. you can pull that out of the configuration system and turn it into whatever you want.

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -163,6 +185,51 @@ void snc_quota_manager::record_response_tp(
+    }
+    ss::steady_clock_type::time_point arm_until
+      = _balancer_timer_last_ran + get_quota_balancer_node_period();
+    const auto now = ss::steady_clock_type::now();
```
> it is also a steady clock

Ha, I have missed that fact. If it is steady, then definitely I can.

> if your code is sensitive to the clock resolution

`task_quota_ms` resolution is good enough for it.

dotnwat replied:
great to hear this!

```
\brief Low-resolution and efficient steady clock
```

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -163,6 +185,51 @@ void snc_quota_manager::record_response_tp(
+        // occurs when balancer takes to run longer than the balancer period,
+        // and on the first iteration
+        if (
+          _balancer_timer_last_ran.time_since_epoch()
```
I don't think it's worth that... for just one condition that controls logging that is true only once in process lifetime. If it was a part of logic than it was definitely worth that.

dotnwat replied:
>         // occurs when balancer takes to run longer than the balancer period,
>        // and on the first iteration

the comment above suggests that this is not something that occurs just once in the process lifetime. comparing against a default constructed time_point might be cleaner than using time_since_epoch == 0 and then you can write a comment like "skip printing until loop has run once" -- but now i'm confused if it is run once or not based on the existing comment.

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -86,12 +88,32 @@ snc_quota_manager::snc_quota_manager()
 
-ss::future<> snc_quota_manager::stop() { return ss::make_ready_future<>(); }
+ss::future<> snc_quota_manager::stop() {
+    if (ss::this_shard_id() == quota_balancer_shard) {
+        _balancer_timer.cancel();
+        return _balancer_gate.close();
```
great catch, I have overlooked that! Apparently it's a pure race why I haven't had segfaults yet.

dotnwat replied:
checking the gate before arming means you should be able to remove the gate check in the balancer step?

dlex replied:
why? Balancer step happens after the timer expires and probabaly after the mutex is released too, so it's a very different point in time anyway. I think it's worth checking there too.

dotnwat replied:
because the shutdown sequence of the entire service will cancel the timer and then wait on the gate via quota_manager::stop.

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -163,6 +185,51 @@ void snc_quota_manager::record_response_tp(
+
+ss::future<> snc_quota_manager::quota_balancer_step() {
+    const auto sem_units = co_await ss::get_units(_balancer_semaphore, 1);
+    _balancer_gate.check();
```
The `.then` continuation should be `.finally` and then the issue will manifest on each application stop. This `check()` here is exclusively to stop the balancer fiber faster in case it is unblocked from semaphore/mutex on shutdown.

dlex commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -70,8 +71,36 @@ class snc_quota_manager
+    /// regardless of what is there now.
+    struct collect_dispense_amount_t {
+        snc_quota_manager::quota_t amount{0};
+        enum { delta, value } amount_is{delta};
```
Then every mentioning of the `enum class` members would require one more identifier in it while not improving anything really: the enum members are currently in `collect_dispense_amount_t` with only 2 more identifiers in it.

dotnwat replied:
>while not improving anything really

strong types eliminate entire classes of mistakes

dlex commented on src/v/kafka/server/snc_quota_manager.h:
```diff
@@ -70,8 +71,36 @@ class snc_quota_manager
+        snc_quota_manager::quota_t amount{0};
+        enum { delta, value } amount_is{delta};
+        bool empty() const noexcept {
+            return amount == 0 && amount_is == delta;
```
That would look like a type name then... IMO right now it reads nicely: "amount equals 0 and the amount is 'delta'"

dotnwat replied:
ok your call. but use an enum class so prevent amount/delta from being compared, for example.

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -163,6 +185,51 @@ void snc_quota_manager::record_response_tp(
+}
+
+ss::future<> snc_quota_manager::quota_balancer_step() {
+    const auto sem_units = co_await ss::get_units(_balancer_semaphore, 1);
```
Moved to a later commit now.

It is protecting from concurrent execution of `quota_balancer_update()`s, as well as concurrent execution of `quota_balancer_update()` and `quota_balancer_step()`.

dotnwat replied:
thanks. i should have come back to update my comment after i made it through the later commits in the PR.

dotnwat commented:
> I think I will leave the semaphore/mutex for now -- if you don't mind -- and consider an alternative later on.

of course. feel free to leave the mutex in place.

but later i'd like to look at this to understand why removing it would result in something that is more complicated.

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -163,6 +185,51 @@ void snc_quota_manager::record_response_tp(
+    // zero period in config means do not run balancer
+    if (v == std::chrono::milliseconds::zero()) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(
+          ss::steady_clock_type::duration::max() / 2);
```
thanks for the explanation. i'm not sure its reasonable to think someone reading this would derive all of this information. please add this information into the code or change the code so that the semantics are clear.

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -163,6 +185,51 @@ void snc_quota_manager::record_response_tp(
+    return v;
+}
+
+void snc_quota_manager::maybe_arm_balancer_timer() {
+    static constexpr auto min_sleep_time = 2ms; // TBD config?
+
+    if (_balancer_timer.armed()) {
+        return;
+    }
```
I've kept the `maybe_` because now it's conditional on the gate state.

dotnwat replied:
ok

dlex commented:
Force push: dozens of review feedback notes addressed, most notably:
* fixed the issue caused by potentially arming balancer timer after the call to `stop()` has been completed
* timer clock changed to `ss::lowres_clock`
* replaced `vasserts` with log
* multiple renames

dlex commented:
Force push: rebased upon the upstream

dlex commented:
CI Failure in ducktape-build-release-clang-amd64: #8179

dotnwat reviewed (commented):
partial review up to commit `config: existing limit data types made signed`

dlex commented on src/v/utils/bottomless_token_bucket.cc:
```diff
@@ -66,3 +71,22 @@ void bottomless_token_bucket::refill(const clock::time_point now) noexcept {
+      v._width.count(),
+      v._tokens,
+      v._burst,
+      v._last_check.time_since_epoch().count());
```
When I look at the log with the token bucket contents dumped, looking at that field I can reason about whether it has changed since the last time or not. That was the main reason to get it there. I haven't done anything more complicated like 2. or 3. to keep it simple and avoid paying performance price for unclear reason.

dotnwat replied:
> to keep it simple

it's not simple now because last_check has no relation to wall-clock time but we are printing out its value relative to the epoch. what other options are available?

> paying performance price

that would need to be judged on a case by case basis. some solutions are unlikely to introduce anything of concern.

dlex replied:
> it's not simple now because last_check has no relation to wall-clock time but we are printing out its value relative to the epoch

and that is good enough for the purpose of seeing if it has changed or not, and also comparing relative changes

> what other options are available?

With regard to what is printed, not many. We can't convert it to wall clock and therefore print a human comprehensible timestamp. To use low-resolution system clock with the sole purpose of nice log printout is an absolute overkill. The only option here is to display duration since a certain point in time, like
* the clock epoch (done now)
* another fixed timepoint (eg 2023-01-01)
* a not-that-fixed timepoint like process start time
  * or like the class object creation time
* the last call to the formatter

IMO none of those options help in the task of seeing the change in the time point, they only can lessen the number of digits in the output.

In terms of how it is printed, there is an option is to represent the time since (epoch or whatever) in human friendly time units, like seconds or HH:MM:SS. This does not add value to that information, but makes it easier to reason about. I think this is something I can do.

dotnwat replied:
> and that is good enough for the purpose of seeing if it has changed or not, and also comparing relative changes

it's fine to log something just to have visibility in what is going on, so i agree that it is good enough.

what is problematic is the the line of code. you could add a comment to fix this, such as:

```
// computing time_since_epoch here has no meaning, we just care about seeing this value change
```

or, you could do something like this

```
// we only care about how this value is changing since it is a steady clock
std::chrono::duration_cast<std::chrono::milliseconds>(v._last_check - clock::time_point::min()).count();
```

which tbh is probably exactly the same thing, but it expresses what is going on more clearly imo.

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -19,6 +19,8 @@ using namespace std::chrono_literals;
 
```
> k/quotas: throughput limits bindings and quotas as inoutpairs
Use inoutpair to represent TP limit property binding, and the quota
token buckets.

commit message needs updated for type name change

dotnwat commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -0,0 +1,703 @@
+              delta,
+              total_surplus,
+              surplus);
+        }
```
would this be a case where you'd want to return early instead of carrying on, since before IIRC it was a vassert(false) situation?

EDIT: oh i see that the effect of the remaining conditional is that the error condition becomes a noop. consider adding a `return` after the vlog to drive that point across.

dotnwat reviewed (commented):
@dlex this looks really great. i have mostly just minor feedback and a couple places that look concerning.

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -0,0 +1,703 @@
+    const ingress_egress_state<quota_t> current_rate = {
+      .in = _shard_quota.in.get_current_rate(),
+      .eg = _shard_quota.eg.get_current_rate()};
+    return std::max(current_rate - _shard_quota_minimum, {0, 0});
```
one more great catch. This needs per component expressions. Btw this is the kind of error I was tryin to avoid with the late `to_each`.

dotnwat replied:
does this mean that `operator<` can be removed?

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -0,0 +1,703 @@
+    const auto units = co_await _balancer_mx.get_units();
+    _balancer_gate.check();
+    _balancer_timer_last_ran = ss::lowres_clock::now();
+    vlog(klog.trace, "qb - Step");
```
I've left this for time context only, probably that can be removed. There's not much of other context indeed.

dotnwat replied:
your call, just a thought i had when reading

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -0,0 +1,703 @@
+        _shard_quota.in.set_width(v);
+        _shard_quota.eg.set_width(v);
+    });
+    _kafka_quota_balancer_node_period.watch([this] {
+        if (_balancer_timer.cancel()) {
+            maybe_arm_balancer_timer();
+        }
```
well actually on all shards where the timer has been armed (`_balancer_timer.cancel()` returns `true`), so effectively shard0. That needs a comment I think.

dotnwat replied:
or check core-0 and be explicit. makes things easier to maintain (ie no comments to update)

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -0,0 +1,703 @@
+
+void snc_quota_manager::record_request_tp(
+  const size_t request_size, const clock::time_point now) noexcept {
+    _shard_quota.in.use(request_size, now);
```
no, the overflow will occur only at practically unreachable request sizes (EBs)

dlex commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -0,0 +1,703 @@
+    vlog(klog.trace, "qb - Step");
+
+    // determine the borrowers and whether any balancing is needed now
+    const auto now = clock::now();
+    const auto borrowers = co_await container().map(
+      [now](snc_quota_manager& qm) -> borrower_t {
```
excellent point, yes for sure

dlex commented:
Force push:
* the issue with `max(ingress_egress_state)` fixed
* the issue with passing ss::lowres_clock value betweed shards fixed
* removed `snc_quota_manager::get_shard_quota_default()` that was confusing in semantics
* multiple cosmetic improvements

dotnwat commented:
Failure is https://github.com/redpanda-data/redpanda/issues/8179

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -103,12 +105,33 @@ snc_quota_manager::snc_quota_manager()
-ss::future<> snc_quota_manager::stop() { return ss::make_ready_future<>(); }
+ss::future<> snc_quota_manager::stop() {
+    if (ss::this_shard_id() == quota_balancer_shard) {
+        _balancer_timer.cancel();
```
I notice that there is no way to cancel in-progress work for shutdown, usually this is done via `ss::abort_source` and having your loop exit on condition that `request_abort` has been called.

In the worst case without this, the future returned by `stop()` will resolve in as long as it will take to execute one iteration of  `quota_balancer_step()`. Would you know how long this could be in the worst case? I see in previous comments you mention how in high load scenarios it could take a long time for a step to complete, in which case this would hold up shutdown.

dotnwat replied:
@graphcareful do you have a case in mind for which abort_source would speed up shutdown? it's nearly always the case that abort_source is useful in aborting things early like sleep(), semaphore::wait, condition_variable::wait, etc... but i don't recall seeing any of these usages. in the case of this PR we effectively have a simulated sleep with an ss::timer, but snc_quota_manager::stop will stop the timer and close the gate for an equivalent speedy shutdown.

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -299,10 +299,236 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
 ss::future<> snc_quota_manager::quota_balancer_update(
-  ingress_egress_state<dispense_quota_amount> /*shard_quotas_update*/) {
+  ingress_egress_state<dispense_quota_amount> shard_quotas_update) {
+    const auto units = co_await _balancer_mx.get_units();
```
This could indefinitely wait if theres some sort of bug introduced, maybe use the version with the timeout, and have the caller retry on detection of exception? This way at least the behavior can be logged and identified

dlex replied:
There are two scopes waiting on this mutex, each running in a fiber different from the other's, never calling each other from inside the scope even asynchronously, and not waiting on anything except for the futures for cross-shard calls that also follow the aforementioned. I see no way to deadlock this. 

Still I see some space for being cautious, like with any syncronization. I have an async task to re-evaluate the need for the 2nd fiber with Noah; if the other fiber and the mutex stay in the code after that, I will add timeouts on `get_units()` as you propose. Does that sounds ok?

dotnwat replied:
I can't identify any specific scenario that could lead to a deadlock on the mutex. Generally I see the pattern:

1. take mutex
2. perform a bounded amount of work
3. release mutex

I think in order for this to be a concern we need to identify a specific scneario. @graphcareful afaict if any bug is introduce that would do something like throw an exception then the mutex is automatically released. it's unusable at that point, but at least it doesn't deadlock.

i think that @graphcareful might be referring to a specific design pattern in which "things fail" and we bubble this up to a point where it is convenient to reset the state of the system to a known state. i wouldn't be opposed to that, but i'm also not seeing it as a big concern here since the cross-core operations are 100% cpu calculation--very simple and benign.

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -193,6 +220,52 @@ snc_quota_manager::get_quota_balancer_node_period() const {
+          .eg = calc_diff(_node_quota_default.eg, new_node_quota_default.eg)};
+        ssx::spawn_with_gate(
+          _balancer_gate, [this, update = shard_quotas_update] {
+              return quota_balancer_update(update);
```
Is another background fiber necessary? Its not like it can increase the processing throughput rate, seems like its just for when config updates arrive. These can be processed by the main fiber.

dlex replied:
That's right they are for config updates. Since the config updates include as many cross core calls as a balancer step does -- if not more -- it was my assumption there is no need to make the property change notifications wait for that to complete.

However there is a stronger argument for asynchronity here: a balancer step and a config update step should not run concurently, and I suspect we definitely won't appreciate a mutex in the main fiber. The concurrency may lead to various inconsistencies in quota sharing between shards. The problems root from the fact that updates to quotas can be both assignments and increments (or decrements). When assignment and increments are applied concurrently by separate fibers, the result is undefined.

dotnwat replied:
>a balancer step and a config update step should not run concurrently

agree, but this presumes a design in which these steps both launch background fibers, which then need to be synchronized. i believe a design exists in which a configuration update merely triggers an iteration of the reconciliation loop (e.g. an early execution of the balancer step) which handles all cases.

in the end, i think this boils down to a design decision. imo @graphcareful  has a good point: if one fiber can correctly reconcile the state of the system with the state of the configuration, then there is no need for two fibers and thus no need for a mutex.

but it's also entirely valid to design this in such a way that the config change reaction and balancer steps are independent processes, which then need to be executed with mutual exclusion for correctness reasons. they are both valid solutions.

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -0,0 +1,696 @@
+        _shard_quota.eg.set_width(v);
+    });
+    _kafka_quota_balancer_node_period.watch([this] {
+        if (_balancer_timer.cancel()) {
```
`maybe_arm_balance_timer` checks that the `gate` is open, however its possible this could be called before the gate has opened

graphcareful commented on src/v/kafka/server/snc_quota_manager.cc:
```diff
@@ -0,0 +1,696 @@
+}
+
+void snc_quota_manager::maybe_arm_balancer_timer() {
+    if (_balancer_gate.is_closed()) {
```
If the only call to this is within the `ssx::spawn_with_gate` then this is an impossible condition

dlex replied:
Another call to this is from `_kafka_quota_balancer_node_period.watch()`. But I see your point: gate can be checked only where needed, and that's fair.

dlex commented:
Force push: 
* balancer gate is only checked when required

dotnwat commented:
> In general (with this fiber/timer design pattern) do we ever care when the time it takes for one iteration of a fiber exceeds its periodicity interval? I don't think its an explicitly handled condition in other places of the codebase today @dotnwat am I right in assuming this?

generally speaking we can think of designs as either assuming that timers fire with a bounded precision, or not. the difference is that in the later case its often required that the implementation adapt to the flexibility in arrival of timer execution. my interpretation of @dlex's design is that it is flexible wrt to when timers fire, but that it attempts to construct an optimal timer schedule in which timers are armed at a variable frequency based on a function of the desired periodicity and the delay needed to maintain that periodicity based on the recent past of schedule of fired timers.
